torch/csrc/distributed/rpc/tensorpipe_agent.h

```diff
@@ -265,16 +274,51 @@ class TensorPipeAgent : public RpcAgent {
       const std::string& remoteName,
       const Message& message) const;
 
+#ifdef USE_CUDA_NOT_ROCM
+  // An RPC-specific CUDAFuture subclass. It overrides the extractDataPtrs
+  // function to handle and only handle RPC Messages.
+  struct TORCH_CUDA_API RpcCUDAFuture final : at::cuda::CUDAFuture {
+   public:
+    using at::cuda::CUDAFuture::CUDAFuture;
+
+   protected:
+    std::vector<std::reference_wrapper<const at::DataPtr>> extractDataPtrs(
+        const at::IValue& value) override {
+      const auto message = value.toCustomClass<Message>();
+      TORCH_INTERNAL_ASSERT(
+          message, "Passed a non-Message type to RpcCUDAFuture"):
+      std::vector<std::reference_wrapper<const at::DataPtr>> data_ptrs;
+      for (const auto& tensor : message->tensors()) {
+        data_ptrs.emplace_back(tensor.storage().data_ptr());
+      }
+      return data_ptrs;
+    }
+  };
+#else
+#endif
+
   // When a request+response completes, we need to mark the future message as
   // complete. However, if its timeout has already expired, it already has an
   // error set. There is no atomic "test-and-set" way to mark a future complete
   // only if it isn't yet. It does exist for errors (setErrorIfNeeded) but, even
   // then, it ends up printing a log message, which may worry the user. To solve
   // both issues we use a separate atomic flag to know the status of the future.
   struct AtomicJitFuture {
-    std::shared_ptr<JitFuture> jitFuture =
-        std::make_shared<JitFuture>(at::AnyClassType::get());
-    std::atomic_flag isComplete = ATOMIC_FLAG_INIT;
+    AtomicJitFuture(bool noCuda = true) : isComplete(ATOMIC_FLAG_INIT) {
```

Comment:
I don't think it's safe to use `ATOMIC_FLAG_INIT` in this way. The doc says:

> [...] can be used [...] with the statement `std::atomic_flag v = ATOMIC_FLAG_INIT`;. It is unspecified if it can be used with other initialization contexts.

https://en.cppreference.com/w/cpp/atomic/ATOMIC_FLAG_INIT

Reply:
reverted